libyul/tools/interpreter/PureInterpreter.cpp

```diff
@@ -0,0 +1,360 @@
+	{
+		EvaluationResult evalRes = evaluate(*_declaration.value, _declaration.variables.size());
+		if (auto* terminated = std::get_if<ExecutionTerminated>(&evalRes)) return *terminated;
+		values = std::move(std::get<EvaluationOk>(evalRes).values);
+	}
+	else
+	{
+		values.assign(_declaration.variables.size(), 0);
+	}
+
+	solAssert(values.size() == _declaration.variables.size(), "");
+	for (size_t i = 0; i < values.size(); ++i)
+	{
+		YulName varName = _declaration.variables.at(i).name;
+		auto [_, isNew] = m_variables.insert_or_assign(varName, values.at(i));
+		solAssert(isNew, "");
+		m_scope->addDeclaredVariable(varName);
+	}
+	return ExecutionOk { ControlFlowState::Default };
+}
+
+ExecutionResult PureInterpreter::operator()(If const& _if)
+{
+	solAssert(_if.condition, "");
+	EvaluationResult conditionRes = evaluate(*_if.condition, 1);
+	if (auto* terminated = std::get_if<ExecutionTerminated>(&conditionRes)) return *terminated;
+
+	if (std::get<EvaluationOk>(conditionRes).values.at(0) != 0)
+		return (*this)(_if.body);
+	return ExecutionOk { ControlFlowState::Default };
+}
+
+ExecutionResult PureInterpreter::operator()(Switch const& _switch)
+{
+	solAssert(_switch.expression, "");
+	solAssert(!_switch.cases.empty(), "");
+
+	EvaluationResult expressionRes = evaluate(*_switch.expression, 1);
+	if (auto* terminated = std::get_if<ExecutionTerminated>(&expressionRes)) return *terminated;
+
+	u256 val = std::get<EvaluationOk>(expressionRes).values.at(0);
+	for (auto const& c: _switch.cases)
+	{
+		bool caseMatched = false;
+		// Default case has to be last.
+		if (!c.value) caseMatched = true;
+		else
+		{
+			EvaluationResult caseRes = evaluate(*c.value, 1);
+			if (auto* terminated = std::get_if<ExecutionTerminated>(&caseRes)) return *terminated;
+			caseMatched = std::get<EvaluationOk>(caseRes).values.at(0) == val;
+		}
+		if (caseMatched) return (*this)(c.body);
+	}
+	return ExecutionOk { ControlFlowState::Default };
+}
+
+ExecutionResult PureInterpreter::operator()(FunctionDefinition const&)
+{
+	return ExecutionOk{ ControlFlowState::Default };
+}
+
+ExecutionResult PureInterpreter::operator()(ForLoop const& _forLoop)
+{
+	solAssert(_forLoop.condition, "");
+
+	enterScope(_forLoop.pre);
+	ScopeGuard g([this]{ leaveScope(); });
+
+	{
+		ExecutionResult execRes = execute(_forLoop.pre.statements);
+		if (execRes == ExecutionResult(ExecutionOk { ControlFlowState::Leave }))
+			return execRes;
+	}
+	while (true)
+	{
+		{
+			EvaluationResult conditionRes = evaluate(*_forLoop.condition, 1);
+			if (auto* terminated = std::get_if<ExecutionTerminated>(&conditionRes)) return *terminated;
+			if (std::get<EvaluationOk>(conditionRes).values.at(0) == 0)
+				break;
+		}
+
+		// Increment step for each loop iteration for loops with
+		// an empty body and post blocks to prevent a deadlock.
+		if (_forLoop.body.statements.size() == 0 && _forLoop.post.statements.size() == 0)
+			if (auto terminated = incrementStatementStep()) return *terminated;
```

Comment:
Some general notes on style to get that out of the way:
- We always put `return` on a new line.
- Please also avoid abbreviations in names. Things like `fun`, `cnt`, `x`, `g`, `vec`, `res` are hard to read and often ambiguous. TBH, we have a lot of bad names already since the code is old and it's also hard to enforce, so some of that might have been just copied, but when writing new code, please try to avoid them.
- Constructor spacing: `ExecutionOk{ControlFlowState::Default};`
- 4-space indents, including in `.yul` files.
- You don't have to include the empty string as message in asserts. We made it optional at some point, but did not strip it from everywhere. When writing new code, you can omit the message.
- I'd avoid making files hidden (with a leading dot in the name).
- In Python code we use the same style for splitting long calls as in C++. I.e. 
    ```python
        gen_test(
            'sgt',
            param_cnt=2,
            calc=lambda p: u2s(p[0]) > u2s(p[1])
        )
    ```

Reply:
Thank you for commenting on the styling. For this part, I have a question about the formatter.

For C++ code style, I see that there is `.clang-format` configuration included. However, using clang-format changes the code drastically. So I deliberately turned off the formatter. Should I use clang format for the new code, or should I try to adjust the code manually? And also the same question applied to Python code.

The is the only questions. I will address the other comments.

Reply:
Besides the code style, is it OK to use macro for the _terminated_ case? I think that check is repetitive, and because it also `return` immediately, I can only think of macro to clean up the code.

Reply:
> Should I use clang format for the new code, or should I try to adjust the code manually?

No, that's an incomplete config that does not fully reflect our style unfortunately. There is an effort underway to get to a workable clang config, but for the time being we format the code manually.

> And also the same question applied to Python code.

For Python we also generally format manually, but our style is generally very close to the one used by [black](https://github.com/psf/black), so if you use that, it should be fine.

Reply:
I have resolved most of the comments for the styling. Hopefully I did not miss any renaming.

I still have the pending question on using macro.

And also I'm not sure where to put the test generator for the `pure` instructions. My idea was that I want to test the calculation of those instructions across some test, against a different implementation (in this case is my Python test generator). I am not sure if there a better way to do this. For now it seems to me that putting the test generator near the tests is still the best choice. Please give some advice for this.

Reply:
> I still have the pending question on using macro.

Sorry, I saw it, but forgot to answer before moving on to the next comment.

I'm not sure about that one yet. I'm not a fan of hiding control flow like this and it seems like there aren't *that* many places where you check this, so I don't see it as a big problem. It seems quite manageable, especially since the interpreter is the only place where this style is used. Maybe a helper to simplify the check bit would be enough?

```c++
EvaluationResult result = evaluate(_expressionStatement.expression, 0);
if (terminated(result))
	return result;
```

On the other hand, the old interpreter used exceptions and that also hides the flow, so a macro may not be worse than that. I do agree that these checks seem tedious and easy to forget so I wouldn't completely reject the idea. An ideal solution would be something like `?` in Rust, but we don't really have an equivalent in C++. For now, the helper approach seems to me like it might be a good compromise.

> And also I'm not sure where to put the test generator for the `pure` instructions.

Not sure about this one either. We don't have a standard practice for this. We generally put scripts into the `scripts/` directory, but then that would break the connection between the generator and those tests. On the other hand, even if it's alongside the tests, I doubt we'll generally pay attention to it in the long term anyway. That's just what happens with these things - we'll likely end up with manual modifications in these tests over time and you won't be able to safely regenerate them without losing those modifications. It may be best to treat the script more as documentation on how they were initially generated rather than something that we'll commit to maintaining in the repo. As such it might be better to just post the script in a comment in the PR.

Reply:
For the macro part, I'll use macro.

> An ideal solution would be something like ? in Rust, but we don't really have an equivalent in C++.

Well this is one reason why I proposed using boost's `outcome` instead of std::variant. They have well documented [TRY macro](https://www.boost.org/doc/libs/1_75_0/libs/outcome/doc/html/tutorial/essential/result/try.html) that looks almost natural as Rust `?`. If you like this idea then I guess I can use it too.

For the python test generator, I'll put it into the comment for the corresponding test it generated.

Reply:
> Well this is one reason why I proposed using boost's `outcome` instead of std::variant. They have well documented [TRY macro](https://www.boost.org/doc/libs/1_75_0/libs/outcome/doc/html/tutorial/essential/result/try.html) that looks almost natural as Rust `?`. If you like this idea then I guess I can use it too.

Let's try that then. I actually do like the idea of trying something from a well established library for this rather than hand-implementing it. It has a better chance of being recognized as something conventional and less surprising.

> For the python test generator, I'll put it into the comment for the corresponding test it generated.

Regarding that, one alternative solution that came to my mind is that we could have one of the CI jobs (e.g. the one that tests Python scripts, since that's guaranteed to have Python) run the script and fail if changes are detected. With this it would actually be acceptable to have it in the repo, because we could rely on people not editing these tests randomly.

I'd also make the script insert a comment at the beginning of the test, saying that it was automatically generated and how to rerun it.

Reply:
I have changed to boost's outcome. Please take a look.